pkg/tsdb/cloudmonitoring/mql_query.go

```diff
@@ -0,0 +1,351 @@
+	span.SetTag("query", query.Query)
+	span.SetTag("from", tsdbQuery.TimeRange.From)
+	span.SetTag("until", tsdbQuery.TimeRange.To)
+	span.SetTag("datasource_id", e.dsInfo.Id)
+	span.SetTag("org_id", e.dsInfo.OrgId)
+
+	defer span.Finish()
+
+	if err := opentracing.GlobalTracer().Inject(
+		span.Context(),
+		opentracing.HTTPHeaders,
+		opentracing.HTTPHeadersCarrier(req.Header)); err != nil {
+		queryResult.Error = err
+		return queryResult, cloudMonitoringResponse{}, nil
+	}
+
+	res, err := ctxhttp.Do(ctx, e.httpClient, req)
+	if err != nil {
+		queryResult.Error = err
+		return queryResult, cloudMonitoringResponse{}, nil
+	}
+
+	data, err := query.unmarshalResponse(res)
+	if err != nil {
+		queryResult.Error = err
+		return queryResult, cloudMonitoringResponse{}, nil
+	}
+
+	return queryResult, data, nil
+}
+
+func (query cloudMonitoringMqlQuery) unmarshalResponse(res *http.Response) (cloudMonitoringResponse, error) {
+	body, err := ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	if err != nil {
+		return cloudMonitoringResponse{}, err
+	}
+
+	if res.StatusCode/100 != 2 {
+		slog.Error("Request failed", "status", res.Status, "body", string(body))
+		return cloudMonitoringResponse{}, fmt.Errorf(string(body))
+	}
+
+	var data cloudMonitoringResponse
+	err = json.Unmarshal(body, &data)
+	if err != nil {
+		slog.Error("Failed to unmarshal CloudMonitoring response", "error", err, "status", res.Status, "body", string(body))
+		return cloudMonitoringResponse{}, err
+	}
+
+	return data, nil
+}
+
+func (query cloudMonitoringMqlQuery) parseResponse(queryRes *tsdb.QueryResult, data cloudMonitoringResponse) error {
+	labels := make(map[string]map[string]bool)
+
+	for _, series := range data.TimeSeriesData {
+		seriesLabels := make(map[string]string)
+		defaultMetricName := ""
+
+		for n, d := range data.TimeSeriesDescriptor.LabelDescriptors {
+			key := toSnakeCase(d.Key)
+			if _, ok := labels[key]; !ok {
+				labels[key] = map[string]bool{}
+			}
+
+			labelValue := series.LabelValues[n]
+			switch d.ValueType {
+			case "BOOL":
+				strVal := strconv.FormatBool(labelValue.BoolValue)
+				labels[key][strVal] = true
+				seriesLabels[key] = strVal
+			case "INT64":
+				intVal := strconv.FormatInt(labelValue.Int64Value, 10)
+				labels[key][intVal] = true
+				seriesLabels[key] = intVal
+			default:
+				labels[key][labelValue.StringValue] = true
+				seriesLabels[key] = labelValue.StringValue
+			}
+		}
+
+		for n, d := range data.TimeSeriesDescriptor.PointDescriptors {
+			if d.ValueType != "DISTRIBUTION" {
+				points := make([]tsdb.TimePoint, 0)
+				// reverse the order to be ascending
+				for i := len(series.PointData) - 1; i >= 0; i-- {
+					point := series.PointData[i]
+					value := point.Values[n].DoubleValue
+
+					if d.ValueType == "INT64" {
+						parsedValue, err := strconv.ParseFloat(point.Values[n].Int64Value, 64)
+						if err == nil {
+							value = parsedValue
+						}
+					}
+
+					if d.ValueType == "BOOL" {
```

Comment:
An `else` makes the relation between the `ifs` clearer:
```suggestion
					} else if d.ValueType == "BOOL" {
```